---
title: "Stat 115 Lab 3"
subtitle: "Batch Effect, Clustering, and DAVID"
author: "Andy Shi"
date: "February 6-8, 2018"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Install and Load Packages

```{r bioc, eval = FALSE}
# install packages from bioconductor
source("https://bioconductor.org/biocLite.R") 
#biocLite("sva")
biocLite("bladderbatch") # for the example data
# etc.
```

```{r libraries, warning = FALSE, message = FALSE}
library(sva)
library(bladderbatch)
library(limma)
```

## Load Data

- Gene expression data from investigation into bladder cancer.
- Outcome: finding differentially expressed genes that are
associated with cancer status (0/1 in the variable `hasCancer`).

```{r loaddata}
data(bladderdata)
pheno <- pData(bladderEset)
pheno$hasCancer <- as.numeric(pheno$cancer == "Cancer")
edata <- exprs(bladderEset)
```

## Clustering

- Use the `hclust` command.
- Can plot the clustering result with `plot`.
- Question: What type of data does `hclust` accept? Hint: `?hclust`

## Clustering Code

```{r clustering}
dist_mat = dist(t(edata)) # clusters by distance between rows of data in a matrix (want samples to be rows)
clustering = hclust(dist_mat, method="complete")
plot(clustering, labels=pheno$batch) # see there is batch effect

```


## More Clustering Plots

- Can you produce a plot where the labels are by cancer status `pheno$cancer`?

## More Clustering Plots: Solution

```{r, clustering2}
plot(clustering, labels=pheno$cancer)
```

## Model Matrix

- Need to specify the model for LIMMA and a null model for ComBat
- If you had other variables you wanted to adjust for, you would put
them in the model for ComBat, too. ComBat removes batch effect.

```{r model-matrix}
model1 <- model.matrix(~hasCancer, data = pheno)
model0 <- model.matrix(~1, data = pheno)
```

## ComBat

- Perform batch effect adjustment

```{r combat}
combat_edata <- ComBat(dat = edata, batch = pheno$batch, mod = model0)
# model matrix removes variables you want to remove the effect of, here it is just "intercept = 1"" variable
```

## Check Clustering Again

- Can you perform clustering to check whether ComBat removed the batch
effect?

## Check Clustering Again: Solution

```{r combat-clusters}
dist_mat = dist(t(combat_edata)) # clusters by distance between rows of data in a matrix (want samples to be rows)
clustering2 = hclust(dist_mat, method="complete")
plot(clustering2, labels=pheno$batch)
# After batch removal you want to draw a plot to make sure the batch effect is removed
# After batch removal, the diseased will cluster together and non-diseased will cluster together
```

## LIMMA

- Your turn: Run LIMMA on the data obtained after running ComBat.
- Find all the differentially expressed genes with adjusted p-value $<= 0.05$ and at least 2 fold change.
- How many genes did you get? Hint: `dim`.

## LIMMA: Solution

- Wrote the row names to a file so we can run DAVID.

```{r limma}
fit <- lmFit(combat_edata, model1)
fit <- eBayes(fit)
tophits = topTable(fit, number=Inf, coef="hasCancer", p.value=0.05, lfc=log2(2))
dim(tophits) # 280 genes

```

## DAVID Prep

- Need to paste a list of genes of interest
- I've found the easiest way to do this is to write the row names of the result of `topTable` to a CSV file, open it in Excel, then copy-paste.

```{r david-prep, eval = FALSE}
write.csv(row.names(tophits), file = "for_david.csv", row.names = FALSE)
```
